recirq/toric_code/toric_code_plaquettes.py

```diff
@@ -0,0 +1,132 @@
+        )
+
+    @classmethod
+    def for_uniform_parity(
+        cls, code: tcr.ToricCodeRectangle, x_value: float, z_value: float
+    ) -> "ToricCodePlaquettes":
+        """Create plaquettes with uniform X and Z expectation values."""
+        x_plaquettes = {(row, col): x_value for row, col in code.x_plaquette_indices()}
+        z_plaquettes = {(row, col): z_value for row, col in code.z_plaquette_indices()}
+        return cls(code, x_plaquettes, z_plaquettes)
+
+    @classmethod
+    def from_global_measurements(
+        cls, code: tcr.ToricCodeRectangle, x_data: pd.DataFrame, z_data: pd.DataFrame
+    ) -> "ToricCodePlaquettes":
+        """Compute stabilizer expectation values from global measurement data.
+
+        Args:
+            code: Toric code rectangle whose plaquette expectation values we store
+            x_data: Result of global measurements in the X basis. DataFrame with a single column of
+                integer values, one for each measurement outcome, as in cirq.Result.data.
+            z_data: Result of global measurements in the Z basis, similar to x_data.
+
+        Returns:
+            ToricCodePlaquettes with expectation values calculated from data
+        """
+        x_plaquettes: Dict[Tuple[int, int], float] = {
+            (row, col): cls.expectation_value(code, x_data, row, col, x_basis=True)
+            for row, col in code.x_plaquette_indices()
+        }
+        z_plaquettes: Dict[Tuple[int, int], float] = {
+            (row, col): cls.expectation_value(code, z_data, row, col, x_basis=False)
+            for row, col in code.z_plaquette_indices()
+        }
+        return cls(code, x_plaquettes, z_plaquettes)
+
+    @classmethod
+    def expectation_value(
+        cls,
+        code: tcr.ToricCodeRectangle,
+        data: pd.DataFrame,
+        row: int,
+        col: int,
+        x_basis: bool,
+    ) -> float:
+        """Compute an expectation value for a single X or Z plaquette.
+
+        Args:
+            code: Toric code rectangle with this plaquette
+            data: DataFrame with a single column of integer values, one for each measurement
+                outcome, as in cirq.Result.data
+            row: Plaquette row
+            col: Plaquette column
+            x_basis: If True, look at the X plaquette at (row, col); otherwise, look at the Z
+                plaquette at (row, col)
+
+        Returns:
+            Expectation value of the plaquette, between -1 and 1
+        """
+        if x_basis:
+            qubit_idxs = code.x_plaquette_to_qubit_idxs(row, col)
+        else:
+            qubit_idxs = code.z_plaquette_to_qubit_idxs(row, col)
+
+        total_qubits = len(code.qubits)
+        parities = data.applymap(
+            lambda value: cls.compute_parity(value, qubit_idxs, total_qubits)
+        )
+        return float(parities.mean())
+
+    @staticmethod
+    def compute_parity(value: int, qubit_idxs: Iterable[int], total_qubits: int) -> int:
```

Comment:
[optional] this whole function seems kindof wild: instead of using `cirq.Result.data`, use `cirq.Result.measurements` then you can do np.logical_xor(bistrings, axis=1) with massive speed boosts!

Reply:
I am going to skip this (optional) suggestion for now, as it would modify the input parameters and change things around a bit.  We can revisit it later if needed.